### PR TITLE
add npu c_barrier op

### PR DESCRIPTION
**Pull request by Baibaifan**
### PR types

New features
### PR changes

OPs
### Describe

add npu c_barrier op. 
This operator is used to balance the waiting between collective communications.
exp:
process1:  sleep(10)->c_barrier_op ->same time to stop.
process2:  c_barrier_op                    ->same time to stop.

process1:
image

process2:
image

### CODE REVIEW

File: paddle/fluid/operators/collective/c_barrier_op.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  using framework::OperatorWithKernel::OperatorWithKernel;
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X", "C_Barrier");
+  }
+};
+
+class CBarrierOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() {
+    AddInput("X", "(Tensor) Input data used in NPUKernel.");
```
Comment (void-main) [inline paddle/fluid/operators/collective/c_barrier_op.cc:33]: the barrier op doesn't need an input, is it required by the framework?

File: paddle/fluid/operators/collective/c_barrier_op_npu.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#endif
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+namespace plat = paddle::platform;
+
+REGISTER_OP_NPU_KERNEL(c_barrier, ops::CBarrierOpNPUKernel<int8_t>,
```
Comment (void-main) [inline paddle/fluid/operators/collective/c_barrier_op_npu.cc:63]: do we really need so many types for barrier op?

File: paddle/fluid/operators/collective/c_barrier_op_npu_test.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  // run
+  f::AttributeMap attrs;
+  attrs["tag"] = std::string("tagx_" + std::to_string(iter));
+  attrs["ring_id"] = 0;
+
+  if (use_barrier) {
+    f::AttributeMap attrs_barrier;
+    attrs_barrier["ring_id"] = 0;
+    attrs_barrier["use_calc_stream"] = use_calc;
+    auto op_barrier = f::OpRegistry::CreateOp("c_barrier", {{"X", {"Data"}}},
```
Comment (void-main) [inline paddle/fluid/operators/collective/c_barrier_op_npu_test.cc]: how could we check the barrier ***is really executed***?

**Review by void-main — COMMENTED** [review]
Overall LGTM

File: paddle/fluid/operators/collective/c_barrier_op.cc (hunk 1 above)
Comment (Baibaifan) [reply paddle/fluid/operators/collective/c_barrier_op.cc:33]: Prevent gc deleting vars

File: paddle/fluid/operators/collective/c_barrier_op_npu.cc (hunk 2 above)
Comment (Baibaifan) [reply paddle/fluid/operators/collective/c_barrier_op_npu.cc:63]: need to use input to register op kernel, so need those types.

File: paddle/fluid/operators/collective/c_barrier_op_npu_test.cc (hunk 3 above)
Comment (Baibaifan) [reply paddle/fluid/operators/collective/c_barrier_op_npu_test.cc]: Add a example, at different time to start, and keep same time to end.

**Review by TCChenlong — APPROVED** [review]

**Review by CheQiXiao — APPROVED** [review]

**Review by PangHua — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Pull request closed without merging**